tensorflow/core/kernels/data/matching_files_dataset_op.cc

```diff
@@ -0,0 +1,305 @@
+        while (!filepath_queue_.empty()) {
+          string cur_dir = filepath_queue_.top();
+          filepath_queue_.pop();
+          std::vector<string> children;
+          Status s = fs->GetChildren(cur_dir, &children);
+          ret.Update(s);
+
+          // If cur_dir has no children, there will two possible situations: 1)
+          // the cur_dir is an empty dir; 2) the cur_dir is actual a file
+          // instead of a director. For the first one, continue to search the
+          // heap. For the second one, if the file matches the pattern, add
+          // it to the heap and finish the search; otherwise, continue the next
+          // search.
+          if (children.empty()) {
+            if (ctx->env()->MatchPath(cur_dir, eval_pattern)) {
+              filepath_queue_.push(cur_dir);
+              return ret;
+            } else {
+              continue;
+            }
+          }
+
+          std::map<string, Status> children_dir_status;
+          // This IsDirectory call can be expensive for some FS. Parallelizing
+          // it.
+          ForEach(
+              ctx, 0, children.size(),
+              [fs, &cur_dir, &children, &fixed_prefix,
+               &children_dir_status](int i) {
+                const string child_path = io::JoinPath(cur_dir, children[i]);
+                // In case the child_path doesn't start with the fixed_prefix,
+                // then we don't need to explore this path.
+                if (!str_util::StartsWith(child_path, fixed_prefix)) {
+                  children_dir_status[child_path] = Status(
+                      tensorflow::error::CANCELLED, "Operation not needed");
+                } else {
+                  children_dir_status[child_path] = fs->IsDirectory(child_path);
+                }
+              });
+
+          for (const auto& child : children) {
+            const string child_dir_path = io::JoinPath(cur_dir, child);
+            const Status child_dir_status = children_dir_status[child];
+            // If the IsDirectory call was cancelled we bail.
+            if (child_dir_status.code() == tensorflow::error::CANCELLED) {
+              continue;
+            }
+
+            if (child_dir_status.ok()) {
+              // push the child dir for next search
+              filepath_queue_.push(child_dir_path);
+            } else {
+              // This case will be a file: if the file matches the pattern, push
+              // it to the heap; otherwise, ignore it.
+              if (ctx->env()->MatchPath(child_dir_path, eval_pattern)) {
+                filepath_queue_.push(child_dir_path);
+              }
+            }
+          }
+        }
+        return ret;
+      }
+
+      static void ForEach(IteratorContext* ctx, int first, int last,
+                          const std::function<void(int)>& f) {
+        for (int i = first; i < last; i++) {
+          (*ctx->runner())([f, i] { std::bind(f, i); });
```

Comment:
should not this be `(*ctx->runner())([f, i] { f(i); });`?

Reply:
If using `(*ctx->runner())([f, i] { f(i); });`, there will an error in the test.log file as below:
```
Segmentation fault: 11
```

So, I change it to be `(*ctx->runner())([f, i] { std::bind(f, i); });`.

Reply:
That is because `ctx->runner()` may in general execute the function you pass to it asynchronously. So by the time the function you passed to it is executed the reference to `f` might no longer be valid (using `std::bind` will create a copy). This made me realize that your code currently does not wait for all of the scheduled function to finish completing, but it should.

I suggest the following refactoring of your code (inlining `ForEach` at its only callsite):

```
auto is_directory_fn = [fs, &cur_dir, &children, &fixed_prefix, &children_dir_status](int i) {
  const string child_path = io::JoinPath(cur_dir, children[i]);
  // In case the child_path doesn't start with the fixed_prefix, then we don't need to explore this path.
  if (!str_util::StartsWith(child_path, fixed_prefix)) {
    children_dir_status[child_path] = Status(
      tensorflow::error::CANCELLED, "Operation not needed");
    } else {
      children_dir_status[child_path] = fs->IsDirectory(child_path);
    }
  });
BlockingCounter counter(children.size());
for (int i = 0; i < children.size(); ++) {
  (*ctx->runner())([&is_directory_fn, &counter, i] {
    is_directory_fn(i);
    counter.DecrementCount();
  }
}
counter.Wait();
```

Reply:
Thanks for your suggestion! It helps me get a better understanding of `ctx-runner()`. 

After updating the code, I got the following error when running the test:
```
*** Received signal 11 ***
*** BEGIN MANGLED STACK TRACE ***
0   libtensorflow_framework.so          0x0000000109fd7d85 _ZN10tensorflow7testingL17StacktraceHandlerEiP9__siginfoPv + 181
1   libsystem_platform.dylib            0x00007fff6c20ff5a _sigtramp + 26
2   libtensorflow_framework.so          0x0000000109d41020 libtensorflow_framework.so + 32
3   _pywrap_tensorflow_internal.so      0x00000001122d5993 _ZNSt3__110__function6__funcIZN10tensorflow4data12_GLOBAL__N_122MatchingFilesDatasetOp7Dataset8Iterator14UpdateIteratorEPNS3_15IteratorContextERKNS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEEEUlvE_NSD_ISI_EEFvvEEclEv + 403
4   libtensorflow_framework.so          0x0000000109fb5fde _ZN5Eigen26NonBlockingThreadPoolTemplIN10tensorflow6thread16EigenEnvironmentEE10WorkerLoopEi + 1342
5   libtensorflow_framework.so          0x0000000109fb598f _ZNSt3__110__function6__funcIZN10tensorflow6thread16EigenEnvironment12CreateThreadENS_8functionIFvvEEEEUlvE_NS_9allocatorIS8_EES6_EclEv + 47
6   libtensorflow_framework.so          0x0000000109fd95b0 _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEENS_8functionIFvvEEEEEEEEPvSB_ + 48
7   libsystem_pthread.dylib             0x00007fff6c219661 _pthread_body + 340
8   libsystem_pthread.dylib             0x00007fff6c21950d _pthread_body + 0
9   libsystem_pthread.dylib             0x00007fff6c218bf9 thread_start + 13
*** END MANGLED STACK TRACE ***

*** Begin stack trace ***
        tensorflow::CurrentStackTrace()
        tensorflow::testing::StacktraceHandler(int, __siginfo*, void*)
        _sigtramp

        std::__1::__function::__func<tensorflow::data::(anonymous namespace)::MatchingFilesDatasetOp::Dataset::Iterator::UpdateIterator(tensorflow::data::IteratorContext*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::'lambda'(), std::__1::allocator<tensorflow::data::(anonymous namespace)::MatchingFilesDatasetOp::Dataset::Iterator::UpdateIterator(tensorflow::data::IteratorContext*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::'lambda'()>, void ()>::operator()()
        Eigen::NonBlockingThreadPoolTempl<tensorflow::thread::EigenEnvironment>::WorkerLoop(int)
        std::__1::__function::__func<tensorflow::thread::EigenEnvironment::CreateThread(std::__1::function<void ()>)::'lambda'(), std::__1::allocator<tensorflow::thread::EigenEnvironment::CreateThread(std::__1::function<void ()>)::'lambda'()>, void ()>::operator()()
        void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::function<void ()> > >(void*)
        _pthread_body
        _pthread_body
        thread_start
*** End stack trace ***
Abort trap: 6
```

It seems to be caused by `children_dir_status`, which is defined by `std::map<string, Status> children_dir_status;` and not thread-safe.

In order to solve this problem, I change the defination of `children_dir_status` to be as below. 
```C++
std::vector<Status> children_dir_status;
children_dir_status.resize(children.size());
```
Then the test cases pass without error. I paste the code here for your checking. I'm wondering if my understanding of the error is right?

```C++
          std::vector<Status> children_dir_status;
          children_dir_status.resize(children.size());

          // This IsDirectory call can be expensive for some FS. Parallelizing
          // it.
          auto is_directory_fn = [fs, &cur_dir, &children, &fixed_prefix,
              &children_dir_status](int i) {
            const string child_path = io::JoinPath(cur_dir, children[i]);
            // In case the child_path doesn't start with the fixed_prefix, then
            // we don't need to explore this path.
            if (!str_util::StartsWith(child_path, fixed_prefix)) {
              children_dir_status[i] = Status(
                  tensorflow::error::CANCELLED, "Operation not needed");
            } else {
              children_dir_status[i] = fs->IsDirectory(child_path);
            }
          };

          BlockingCounter counter(children.size());
          for (int i = 0; i < children.size(); i++) {
            (*ctx->runner())([&is_directory_fn, &counter, i] {
              is_directory_fn(i);
              counter.DecrementCount();
            });
          }
          counter.Wait();

          for (int i = 0; i < children.size(); i++) {
            const string child_dir_path = io::JoinPath(cur_dir, children[i]);
            const Status child_dir_status = children_dir_status[i];
            // If the IsDirectory call was cancelled we bail.
            if (child_dir_status.code() == tensorflow::error::CANCELLED) {
              continue;
            }

            if (child_dir_status.ok()) {
              // push the child dir for next search
              filepath_queue_.push(child_dir_path);
            } else {
              // This case will be a file: if the file matches the pattern, push
              // it to the heap; otherwise, ignore it.
              if (ctx->env()->MatchPath(child_dir_path, eval_pattern)) {
                filepath_queue_.push(child_dir_path);
              }
            }
          }
```